File: fixtures/rsi/desktop-product/tasks.py
"""Native Goal/patch/Jobs mechanism checks, with explicit provider response gates."""
import json
import re
import threading


class ProviderControl:
    def __init__(self):
        self.lock = threading.Lock()
        self.held = {}
        self.sequence = 0

    def handle(self, handler, body):
        messages = body.get('messages', [])
        index = next((i for i in range(len(messages) - 1, -1, -1) if messages[i]['role'] == 'user'), -1)
        content = messages[index].get('content', '') if index >= 0 else ''
        prompt = content if isinstance(content, str) else '\n'.join(item.get('text', '') for item in content)
        if not any(marker in prompt for marker in ('Native inline patch', 'Native Goal hold', 'Native background job')):
            return False
        completed = any(message['role'] == 'tool' for message in messages[index + 1:])
        with self.lock:
            self.sequence += 1
            sequence = self.sequence
        handler.send_response(200)
        handler.send_header('Content-Type', 'text/event-stream')
        handler.end_headers()
        def event(delta, reason=None):
            handler.wfile.write(('data: ' + json.dumps({'choices': [{'delta': delta, 'finish_reason': reason}]}) + '\n\n').encode())
            handler.wfile.flush()
        if 'Native Goal hold' in prompt or ('Native background job' in prompt and completed):
            released = threading.Event()
            with self.lock:
                self.held[sequence] = (prompt, released)
            try:
                event({'content': 'Waiting for native fixture release.'})
                if not released.wait(60):
                    raise TimeoutError('native provider gate was not released')
            finally:
                with self.lock:
                    self.held.pop(sequence, None)
        elif not completed:
            if 'Native inline patch' in prompt:
                name, arguments = 'apply_patch', {'patch': '*** Begin Patch\n*** Update File: card.txt\n@@\n-before\n+after · 界\n*** End Patch\n'}
            else:
                name, arguments = 'bash', {'command': "printf 'job-ready\\n'; while [ ! -f job-release ]; do sleep 0.05; done; printf 'job-done\\n'", 'run_in_background': True}
            event({'tool_calls': [{'index': 0, 'id': f'native-call-{sequence}', 'type': 'function',
                                  'function': {'name': name, 'arguments': json.dumps(arguments)}}]})
            event({}, 'tool_calls')
            handler.wfile.write(b'data: [DONE]\n\n')
            return True
        event({'content': 'Native task stream released.'})
        event({}, 'stop')
        handler.wfile.write(b'data: [DONE]\n\n')
        return True

    def release(self, marker):
        with self.lock:
            matches = [event for prompt, event in self.held.values() if marker in prompt]
        assert matches, f'no native provider gate for {marker}'
        for event in matches:
            event.set()

    def close(self):
        with self.lock:
            for _, event in self.held.values():
                event.set()


def verify(script, button, fill, until, screenshot, workspace, report, provider):
    button('Trajectory')
    def text(selector):
        return script('return document.querySelector(arguments[0])?.innerText??""', [selector])
    def has(value):
        return value in text('#detail .ui-contribution')
    def close():
        button('Close details')
        until(lambda: script('return !document.querySelector("#detail").open'))
    def send(prompt, expected):
        fill('textarea[aria-label="Main message"]', prompt)
        button('Send ↗')
        def ready():
            pending = script('return [...document.querySelectorAll(".pending button")].find(b=>b.textContent.startsWith("Review:"))?.textContent??null')
            if pending:
                button(pending)
                button('Allow once')
            return expected in text('.transcript')
        until(ready)
    (workspace / 'card.txt').write_text('before\n')
    send('Native inline patch', 'Native task stream released.')
    until(lambda: '+after · 界' in text('.inline-card:not([hidden])'))
    assert (workspace / 'card.txt').read_text() == 'after · 界\n'
    (workspace / 'card.txt').write_text('later unrelated edit\n')
    assert '-before' in text('.inline-card:not([hidden])') and 'later unrelated edit' not in text('.inline-card:not([hidden])')
    script('document.querySelector(".inline-card:not([hidden]) pre").scrollIntoView({block:"center"})')
    inline_geometry = until(lambda: script('''
        const diff=document.querySelector('.inline-card:not([hidden]) pre'), box=diff.getBoundingClientRect();
        if(box.width<=0||box.height<=0||box.top<0||box.bottom>innerHeight||box.left<0||box.right>innerWidth)return null;
        for(let parent=diff.parentElement;parent;parent=parent.parentElement){
            const style=getComputedStyle(parent),clip=parent.getBoundingClientRect();
            if(/(auto|scroll|hidden|clip)/.test(style.overflowY)&&(box.top<clip.top||box.bottom>clip.bottom))return null;
            if(/(auto|scroll|hidden|clip)/.test(style.overflowX)&&(box.left<clip.left||box.right>clip.right))return null;
        }
        const hit=diff.contains(document.elementFromPoint(box.x+box.width/2,box.y+box.height/2));
        return hit?{top:box.top,bottom:box.bottom,left:box.left,right:box.right,hit}:null;
    '''))
    screenshot('tasks-inline.png')
    button('Goal')
    fill('[aria-label="Goal objective"]', 'Native Goal hold for control evidence')
    fill('[aria-label="Maximum automatic rounds"]', '3')
    button('Create and start Goal')
    until(lambda: has('Allocated rounds: 1 / 3') and has('Current driving: Armed'))
    until(lambda: 'Waiting for native fixture release.' in text('.transcript'))
    screenshot('tasks-goal-armed.png')
    button('Pause after current round')
    until(lambda: has('Durable phase: Paused') and has('Current driving: Disarmed'))
    screenshot('tasks-goal-paused.png')
    provider.release('Native Goal hold')
    until(lambda: has('Driver: Disarmed') and text('.pane-status') == 'Completed')
    button('Resume Goal')
    until(lambda: has('Allocated rounds: 2 / 3') and has('Driver: Waiting'))
    until(lambda: text('.transcript').count('Waiting for native fixture release.') == 2)
    screenshot('tasks-goal-resumed.png')
    button('Cancel automatic round')
    until(lambda: text('.pane-status') == 'Cancelled' and has('Driver: Disarmed'))
    provider.release('Native Goal hold')
    screenshot('tasks-goal-cancelled.png')
    goal_text = text('#detail')
    allocated, cap = map(int, re.search(r'Allocated rounds: (\d+) / (\d+)', goal_text).groups())
    assert (allocated, cap) == (2, 3)
    close()
    send('Native background job', 'Started background Bash job')
    button('Current-Turn Jobs')
    until(lambda: has('Running') and has('Reported: false'))
    screenshot('tasks-jobs-running.png')
    (workspace / 'job-release').write_text('release\n')
    def finished():
        button('Refresh current Turn')
        return has('Completed') and has('Reported: false · Output retained: true')
    until(finished)
    screenshot('tasks-jobs-completed.png')
    provider.release('Native background job')
    until(lambda: text('.pane-status') == 'Failed')
    assert 'background jobs completed without an explicit report: job-1' in text('.transcript')
    jobs_terminal_text = text('.transcript')
    button('Refresh current Turn')
    until(lambda: has('No active Turn.'))
    screenshot('tasks-jobs-revoked.png')
    close()
    (report / 'tasks.json').write_text(json.dumps({'ok': True, 'native_clicks': True,
        'recorded_patch': True, 'inline_geometry': inline_geometry, 'goal_allocations_after_resume': allocated, 'goal_cap': cap,
        'goal_after_cancel_text': goal_text, 'jobs_terminal_text': jobs_terminal_text,
        'jobs_terminal': 'failed_unreported_job'}, indent=2))

File: fixtures/rsi/web-product/tasks.mjs
import assert from "node:assert/strict";
import { mkdir, writeFile, readFile, copyFile, chmod } from "node:fs/promises";
import { join, resolve } from "node:path";
import { createHash } from "node:crypto";
import { chromium, firefox } from "playwright";
import { startService } from "./service.mjs";
import { browserNames, assertControls, assertNoNotices, recordTaskFailure } from "./task-checks.mjs";

const names = browserNames(process.env.RSI_WEB_BROWSER);
const report = process.env.RSI_WEB_REPORT;
const assets = process.env.RSI_WEB_ASSETS;
if (!report || !assets) throw new Error("Set RSI_WEB_REPORT and RSI_WEB_ASSETS");
await mkdir(report, { recursive: false });
const binary = join(report, "rsi");
await copyFile(process.env.RSI_WEB_BINARY ?? resolve("target/debug/rsi"), binary);
await chmod(binary, 0o700);
const binaryHash = createHash("sha256").update(await readFile(binary)).digest("hex");

async function until(predicate, label, maximum = 30000) {
  const deadline = Date.now() + maximum;
  while (!await predicate()) {
    if (Date.now() >= deadline) throw new Error(`${label} timeout`);
    await new Promise(resolve => setTimeout(resolve, 50));
  }
}
async function geometry(page, pane, detail) {
  return page.evaluate(({ detail, pane }) => {
    const root = document.querySelector(detail ? "#detail" : pane);
    const rect = root.getBoundingClientRect();
    const buttons = [...root.querySelectorAll("button")].filter(button => {
      const box = button.getBoundingClientRect();
      if (button.disabled || box.width <= 0 || box.height <= 0 || box.top < 0 || box.bottom > innerHeight) return false;
      for (let parent = button.parentElement; parent; parent = parent.parentElement) {
        const style = getComputedStyle(parent), clip = parent.getBoundingClientRect();
        if (/(auto|scroll|hidden|clip)/.test(style.overflowY) && (box.top < clip.top || box.bottom > clip.bottom)) return false;
        if (/(auto|scroll|hidden|clip)/.test(style.overflowX) && (box.left < clip.left || box.right > clip.right)) return false;
      }
      return true;
    });
    const hit = button => {
      const box = button.getBoundingClientRect();
      return button.contains(document.elementFromPoint(box.x + box.width / 2, box.y + box.height / 2));
    };
    const send = [...root.querySelectorAll("button")].find(button => button.textContent === "Send ↗");
    return { viewport: [innerWidth, innerHeight], page_width: document.documentElement.scrollWidth,
      root: { left: rect.left, right: rect.right, width: rect.width },
      controls: buttons.map(button => ({ label: button.textContent, hit: hit(button) })),
      send_hit: detail ? null : !!send && hit(send) };
  }, { detail, pane });
}

const results = [];
for (const [name, engine] of [["chromium", chromium], ["firefox", firefox]]) {
  if (!names.includes(name)) continue;
  const directory = join(report, name);
  await mkdir(directory);
  const browser = await engine.launch({ headless: true });
  let service, page;
  const errors = [], measurements = [];
  try {
    service = await startService({ binary, assets, report: directory });
    await writeFile(join(service.workspace, "card.txt"), "before\n");
    const context = await browser.newContext({ ignoreHTTPSErrors: true, viewport: { width: 1440, height: 980 } });
    await context.addInitScript(() => {
      const identities = new WeakMap(); let sequence = 0;
      window.taskInputEvents = [];
      window.taskVisibleRequests = [];
      window.taskInvocations = [];
      window.taskGoalErrors = [];
      const observed = new WeakSet();
      const boundedPush = (list, item) => { list.push(item); if (list.length > 64) list.shift(); };
      const observeGoalError = (text, source) => {
        if (!/Goal control rejected|Goal control:|command revision conflict|UI action or surface has retired|Control outcome is unresolved/.test(text)) return;
        const diagnostic = text.slice(0, 4096);
        if (window.taskGoalErrors.at(-1)?.diagnostic !== diagnostic) boundedPush(window.taskGoalErrors, { time: performance.now(), source, diagnostic });
      };
      document.addEventListener("DOMContentLoaded", () => {
        new MutationObserver(() => observeGoalError(document.querySelector("#detail")?.textContent ?? "", "DOM"))
          .observe(document.body, { subtree: true, childList: true, characterData: true });
      });
      const post = Worker.prototype.postMessage;
      Worker.prototype.postMessage = function(message, ...rest) {
        if (!observed.has(this)) {
          observed.add(this);
          this.addEventListener("message", ({ data }) => {
            if (data?.kind === "reply") {
              const invocation = window.taskInvocations.find(item => item.id === data.id);
              if (invocation) invocation.reply = { time: performance.now(), error: data.error?.slice(0, 4096), notAdmitted: data.notAdmitted, result: typeof data.result === "string" ? data.result.slice(0, 4096) : data.result };
            }
            if (data?.kind === "view") {
              const frame = JSON.parse(data.view);
              observeGoalError(JSON.stringify(frame.view?.ui_detail ?? frame.sections?.ui_detail ?? {}), "Worker frame");
            }
          });
        }
        if (message?.method === "command") {
          const request = JSON.parse(message.payload);
          if (request.action === "ui_invoke") {
            const value = request.input?.value;
            boundedPush(window.taskInvocations, { id: message.id, time: performance.now(), ticket: request.ticket, name: request.name,
              kind: value?.kind, request_id: value?.request?.request_id ?? value?.request,
              expected_revision: value?.request?.expected_revision ?? value?.revision,
              action: value?.request?.action, reply: null });
          }
          if (request.action === "ui_visible" && request.keys.length) {
            window.taskVisibleRequests.push(request);
            if (window.taskVisibleRequests.length === 1) {
              queueMicrotask(() => this.dispatchEvent(new MessageEvent("message", { data: {
                kind: "reply", id: message.id, error: "Injected visible-hint admission failure", notAdmitted: true,
              } })));
              return;
            }
          }
        }
        return post.call(this, message, ...rest);
      };
      for (const type of ["mousedown", "mouseup", "click"]) document.addEventListener(type, event => {
        const target = event.target.closest?.("button");
        if (!target) return;
        if (!identities.has(target)) identities.set(target, ++sequence);
        window.taskInputEvents.push({ type, label: target.textContent, id: identities.get(target), time: performance.now() });
        if (window.taskInputEvents.length > 64) window.taskInputEvents.shift();
      }, true);
    });
    page = await context.newPage();
    page.setDefaultTimeout(30000);
    page.on("pageerror", error => errors.push(error.message));
    await page.goto(service.origin);
    await page.getByLabel("Device registration receipt").fill(JSON.stringify(service.register(`${name} task evidence`)));
    await page.getByRole("button", { name: "Connect", exact: true }).click();
    await page.locator("#workbench").waitFor({ state: "visible" });
    await page.locator(".workspace-add summary").click();
    await page.getByLabel("Server directory").fill(service.workspace);
    await page.locator("#workspace-form").getByRole("button", { name: "Add workspace", exact: true }).click();
    await page.locator("#workspaces .nav-item").click();
    await page.getByRole("button", { name: "Trajectory", exact: true }).click();
    const paneSelector = '[aria-label="Main conversation"]';
    const pane = page.locator(paneSelector);
    const detail = page.locator("#detail .ui-contribution");
    const goalUntil = async (predicate, label) => until(async () => {
      const evidence = await page.evaluate(() => ({ errors: window.taskGoalErrors, replies: window.taskInvocations.filter(item => item.name === "goal" && item.reply?.error) }));
      assert.deepEqual(evidence, { errors: [], replies: [] }, `${label}: explicit Goal failure ${JSON.stringify(evidence)}`);
      return predicate();
    }, label);
    const goalVisible = async (locator, label) => goalUntil(() => locator.isVisible(), label);
    const close = async () => {
      if (await page.locator("#detail").isVisible()) await page.getByRole("button", { name: "Close details", exact: true }).click();
      await page.locator("#detail").waitFor({ state: "hidden" });
    };
    const surface = async name => { await close(); await page.getByRole("button", { name, exact: true }).click(); };
    const capture = async (label, isDetail = false, expected = []) => {
      for (const [size, viewport] of [["desktop", { width: 1440, height: 980 }], ["narrow", { width: 390, height: 844 }]]) {
        await page.setViewportSize(viewport);
        await page.evaluate(() => new Promise(resolve => requestAnimationFrame(() => requestAnimationFrame(resolve))));
        const expected_controls = await assertControls(page, isDetail ? detail : pane, expected);
        await assertNoNotices(page);
        const metric = await geometry(page, paneSelector, isDetail);
        assert(metric.page_width <= viewport.width, JSON.stringify(metric));
        assert(metric.root.left >= 0 && metric.root.right <= viewport.width + 1, JSON.stringify(metric));
        assert(metric.controls.length > 0 && metric.controls.every(button => button.hit), JSON.stringify(metric));
        if (!isDetail) assert(metric.send_hit, JSON.stringify(metric));
        measurements.push({ label, size, expected_controls, ...metric });
        await page.screenshot({ path: join(directory, `${label}-${size}.png`), fullPage: true });
      }
      await page.setViewportSize({ width: 1440, height: 980 });
    };
    const send = async text => {
      await close();
      const before = service.provider.requests.length;
      await pane.getByLabel("Main message", { exact: true }).fill(text);
      assert.equal(await pane.getByLabel("Main message", { exact: true }).inputValue(), text);
      await pane.getByRole("button", { name: "Send ↗", exact: true }).click();
      await until(async () => {
        const review = pane.locator(".pending button").filter({ hasText: "Review:" });
        if (await review.count()) {
          await review.first().click();
          await page.getByRole("button", { name: "Allow once", exact: true }).click();
        }
        return service.provider.requests.length > before && (
          /Completed|Failed/.test(await pane.locator(".pane-status").innerText()) ||
          (await pane.locator(".transcript .message.assistant").last().innerText().catch(() => "")).includes("Waiting for fixture release."));
      }, "submitted Turn");
    };

    await send("Please record an inline patch");
    const inline = pane.locator(".inline-card:visible");
    await inline.locator("pre").filter({ hasText: "+after · 界" }).waitFor();
    assert.equal(await readFile(join(service.workspace, "card.txt"), "utf8"), "after · 界\n");
    await writeFile(join(service.workspace, "card.txt"), "unrelated later edit\n");
    assert.match(await inline.innerText(), /-before/);
    assert.doesNotMatch(await inline.innerText(), /unrelated later edit/);
    await capture("recorded-inline-patch", false, ["Complete result"]);
    const visibilityRetry = await page.evaluate(() => window.taskVisibleRequests.slice(0, 2));
    assert.equal(visibilityRetry.length, 2);
    assert.deepEqual(visibilityRetry[0], visibilityRetry[1], "visible-hint retry must retain its sequence and selection");
    await inline.getByRole("button", { name: "Complete result", exact: true }).click();
    await inline.locator("pre").filter({ hasText: '"status": "applied"' }).waitFor();

    await surface("Goal");
    await detail.getByLabel("Goal objective", { exact: true }).fill("UI Goal hold for explicit control evidence");
    await detail.getByLabel("Maximum automatic rounds", { exact: true }).fill("3");
    await capture("goal-create", true, ["Create and start Goal"]);
    const beforeGoal = service.provider.requests.length;
    await detail.getByRole("button", { name: "Create and start Goal", exact: true }).click();
    await detail.locator(".ui-field").filter({ hasText: "Allocated rounds: 1 / 3" }).waitFor();
    await detail.locator(".ui-field").filter({ hasText: "Current driving: Armed" }).waitFor();
    await until(() => service.provider.requests.length === beforeGoal + 1, "Goal stream entered");
    await detail.locator(".ui-field").filter({ hasText: "Driver: Waiting" }).waitFor();
    await capture("goal-armed", true, ["Pause after current round", "Cancel automatic round"]);
    await detail.getByRole("button", { name: "Pause after current round", exact: true }).click();
    await detail.locator(".ui-field").filter({ hasText: "Durable phase: Paused" }).waitFor();
    await detail.locator(".ui-field").filter({ hasText: "Current driving: Disarmed" }).waitFor();
    assert.equal(service.provider.requests.length, beforeGoal + 1);
    await capture("goal-paused-claimed", true, ["Resume Goal", "Cancel automatic round"]);
    service.provider.release("UI Goal hold");
    await pane.locator(".pane-status").filter({ hasText: "Completed" }).waitFor();
    await detail.locator(".ui-field").filter({ hasText: "Driver: Disarmed" }).waitFor();
    await goalVisible(detail.getByRole("button", { name: "Create and start Goal", exact: true }), "Paused first-round settlement projection");
    await detail.getByRole("button", { name: "Resume Goal", exact: true }).click();
    await goalVisible(detail.locator(".ui-field").filter({ hasText: "Allocated rounds: 2 / 3" }), "second Goal allocation");
    await goalUntil(() => service.provider.requests.length === beforeGoal + 2, "resumed Goal stream entered");
    await detail.getByRole("button", { name: "Cancel automatic round", exact: true }).click();
    await pane.locator(".pane-status").filter({ hasText: "Cancelled" }).waitFor();
    await detail.locator(".ui-field").filter({ hasText: "Driver: Disarmed" }).waitFor();
    assert.match(await detail.innerText(), /Allocated rounds: 2 \/ 3/);
    await capture("goal-cancelled", true, ["Resume Goal", "Create and start Goal"]);

    await send("Please observe background job");
    await surface("Current-Turn Jobs");
    await detail.locator("pre").filter({ hasText: "Running" }).waitFor();
    assert.match(await detail.innerText(), /Reported: false/);
    const requestsBeforeRead = service.provider.requests.length;
    await capture("jobs-running", true, ["Refresh current Turn"]);
    await writeFile(join(service.workspace, "job-release"), "release\n");
    await until(async () => {
      await detail.getByRole("button", { name: "Refresh current Turn", exact: true }).click();
      return /Completed/.test(await detail.innerText());
    }, "job completion status");
    assert.match(await detail.innerText(), /Reported: false · Output retained: true/);
    assert.equal(service.provider.requests.length, requestsBeforeRead);
    await capture("jobs-completed-unreported", true, ["Refresh current Turn"]);
    service.provider.release("observe background job");
    await pane.locator(".pane-status").filter({ hasText: "Failed" }).waitFor();
    await pane.locator(".transcript").getByText("background jobs completed without an explicit report: job-1", { exact: true }).waitFor();
    await detail.getByRole("button", { name: "Refresh current Turn", exact: true }).click();
    await detail.getByText("No active Turn. Jobs are unavailable after their Turn finishes.", { exact: true }).waitFor();
    await capture("jobs-revoked", true, ["Refresh current Turn"]);
    await assertNoNotices(page);
    await close();
    await page.locator("#sign-out").click();
    await page.locator("#login").waitFor({ state: "visible" });
    assert.deepEqual(errors, []);
    results.push({ browser: name, version: browser.version(), binary_sha256: binaryHash, ok: true,
      ui_invocations: await page.evaluate(() => window.taskInvocations), goal_errors: await page.evaluate(() => window.taskGoalErrors),
      measurements, visibility_retry: visibilityRetry, requests: service.provider.requests, jobs_terminal: "failed_unreported_job", clean_sign_out: true });
  } catch (error) {
    await recordTaskFailure(page, directory, { error: String(error), errors, measurements, requests: service?.provider.requests,
      ui_invocations: await page?.evaluate(() => window.taskInvocations).catch(() => []),
      goal_errors: await page?.evaluate(() => window.taskGoalErrors).catch(() => []),
      input_events: await page?.evaluate(() => window.taskInputEvents).catch(() => []) }).catch(reportError => { error.cause = reportError; });
    throw error;
  } finally {
    await browser.close();
    await service?.close();
    await writeFile(join(report, "results.json"), JSON.stringify(results, null, 2));
  }
}
assert.equal(results.length, names.length, "every selected browser must complete the task probe");
